### PR TITLE
Check that the property uri is a Uri instance

### PR DESCRIPTION
**Pull request by adriansuter**

### CODE REVIEW

**Comment by adriansuter** [comment]
Pull request for issue #2658.

I decided to check if the property is of type `\Slim\Http\Uri`.

**Comment by l0gicgate** [comment]
Can you write a test case for it please?

File: Slim/Http/Request.php [hunk 1]
```diff
@@ -497,7 +497,11 @@ public function getRequestTarget()
             return '/';
         }
 
-        $basePath = $this->uri->getBasePath();
+        if ($this->uri instanceof Uri) {
+            $basePath = $this->uri->getBasePath();
+        } else {
```
Comment (l0gicgate) [inline Slim/Http/Request.php:502]: Can you add a test case for this else case please?

**Comment by l0gicgate** [comment]
You don't need to create an entire stub to mock that URI. Instead just prophesize the interface:
```php
public function testGetRequestTargetIfUriPsr7Only() {
    $basePath = '/base/path';

    $uriProphecy = $this->prophesize(UriInterface::class);
    $uriProphecy
        ->getBasePath()
        ->willReturn($basePath)
        ->shouldBeCalledOnce();
    
    $request = $this->requestFactory();
    $prop = new ReflectionProperty($request, 'uri');
    $prop->setAccessible(true);
    $prop->setValue($request, $uriProphecy->reveal());
    
    $this->assertEquals($basePath, $request->getRequestTarget());
}
```

Note: In the example above I'm mocking `UriInterface::class` in this test case it should be `Uri::class` since we want to detect that we are being passed a Slim `Uri` object.

**Comment by adriansuter** [comment]
Ah, I didn't know that. But wait, shouldn't we make sure, that the method `getBasePath()` was never called if the uri is only an instance of a class implementing RouterInterface which has no such method.

**Comment by l0gicgate** [comment]
> the uri is only an instance of a class implementing RouterInterface which has no such method

@adriansuter you mean a class implementing `Uri`. You can mock that instead of `UriInterface`. Still the same synthax.

```php
$uriProphecy = $this->prophesize(Uri::class);
```

**Comment by adriansuter** [comment]
No, I mean a class implementing the `UriInterface` directly. That is actually the case for which I have added the `else` statement in `getRequestTarget`.

So in my opinion the test should be, whether the method works in case the uri is **not** of type (or any extending type) of `Uri`. Because that is the case when we can't call `getBasePath()`. Remember that this method is not part of PSR7.

The other case, when the uri **is** a `Uri` (or any extending type) works, as all other tests assure that already. Actually they use `Uri` all the time, as the request factory creates an instance of type `Uri`.

**Comment by l0gicgate** [comment]
**Here's what the two tests should be:**
```php
public function testGetRequestTargetWithSlimPsr7Uri() {
    $basePath = '/base/path';

    $uriProphecy = $this->prophesize(Uri::class);
    $uriProphecy
        ->getBasePath()
        ->willReturn($basePath)
        ->shouldBeCalledOnce();
    
    $request = $this->requestFactory();
    $prop = new ReflectionProperty($request, 'uri');
    $prop->setAccessible(true);
    $prop->setValue($request, $uriProphecy->reveal());
    
    $this->assertEquals($basePath . '/', $request->getRequestTarget());
}
```
```php
public function testGetRequestTargetWithNonSlimPsr7Uri() {
    // We still pass in a UriInterface, which isn't an instance of Slim URI
    $uriProphecy = $this->prophesize(UriInterface::class);
    $uriProphecy
        ->getBasePath()
        ->shouldBeNotBeCalled();
    
    $request = $this->requestFactory();
    $prop = new ReflectionProperty($request, 'uri');
    $prop->setAccessible(true);
    $prop->setValue($request, $uriProphecy->reveal());
    
    $this->assertEquals('/', $request->getRequestTarget());
}
```

Please get rid of that Uri stub.

Thanks

**Comment by adriansuter** [comment]
Thanks.

Okay, you wanted both tests, I see. Of course I will remove the "stub".

(Tomorrow, it's one o'clock in the morning)

**Comment by adriansuter** [comment]
In the second test case `testGetRequestTargetWithNonSlimPsr7Uri()` I could not add a `MethodProphecy` as PHPUnit would immediately return an error because the method `getBasePath()` is not defined in `UriInterface`.

**Review by l0gicgate — APPROVED** [review]

**Pull request merged**